### DISCUSSION TITLE
配置完成但拦截不生效

### DISCUSSION THREAD

**Discussion by lr784521** [wurensen/gradle_plugin_android_aspectjx#43] [Q&A]
classpath 'com.android.tools.build:gradle:7.2.1'
        classpath 'io.github.wurensen:gradle-android-plugin-aspectjx:3.2.0'

apply plugin: 'io.github.wurensen.android-aspectjx'

@Aspect
public class MainUtil {

    //实现全局监听点击事件
//    @Pointcut("execution(* com.huayun.application.page.splash.SplashActivity.aasee())")
    @Pointcut("execution(* android.view.View.OnClickListener.onClick(android.view.View))")
    public void setOnClick(){
    }

    @Around("setOnClick()")
    public void setOnClickListener(ProceedingJoinPoint point){
        LogUtil.xLoge("setOnClickListenersetOnClickListener");
        try {
            point.proceed();
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        }
    }

}

**Comment by wurensen** [comment]
@lr784521 自己检查下织入规则是否正确，正确情况下可以查看下是否参与了织入：[问题排查](https://github.com/wurensen/gradle_plugin_android_aspectjx#%E9%97%AE%E9%A2%98%E6%8E%92%E6%9F%A5)

**Comment by wurensen** [comment]
@lr784521 不对啊，你这规则是对`android.view.View.OnClickListener.onClick`进行织入啊，这是Android framework的类，是没办法直接修改的啊。

**Comment by lr784521** [comment]
![506d032f587ebac31e1187ef012a08f](https://user-images.githubusercontent.com/36124038/194795155-a3378082-4615-4b6b-82ca-f87198bb453e.jpg)
![0b02e88868dd4b7de9c74889bd9a23f](https://user-images.githubusercontent.com/36124038/194795164-b4489062-2be8-486c-91c9-e33c5df9f2fb.jpg)


![6847d626540449ffaf60c2bea29e9b8](https://user-images.githubusercontent.com/36124038/194795350-40bb5aa3-9e59-4283-bd81-6e41f988981e.j
![d25540326948f7fc0ff2a4fbb73affa](https://user-images.githubusercontent.com/36124038/194795356-4840d7be-f786-4ed3-8ddb-1c9fa50f6b86.jpg)
pg)

这是我之前写的demo  用的是
        classpath 'com.hujiang.aspectjx:gradle-android-plugin-aspectjx:2.0.10'
   api 'org.aspectj:aspectjrt:1.9.7'
老的  是可以拦截的

这是我的联系方式  1243418334@qq.com   我希望把demo分享给你

**Comment by lr784521** [comment]
![09b6beb2a0eab5eaadd4b11cacf676b](https://user-images.githubusercontent.com/36124038/194795575-9104c9aa-438e-4d92-b889-ab2775036a25.jpg)
![47fb230a8d89ce2a9aab0bd4bf425cc](https://user-images.githubusercontent.com/36124038/194795579-b081f0db-71c0-4491-90c1-418f72992b9f.jpg)

而且 你看我这个规则 就是拦截 一个普通方法  也没有生效，所以我感觉  是不是 我的配置哪里不对呢，希望您指导下  非常感谢

**Comment by wurensen** [comment]
@lr784521 你到对应`module`的`build/tmp/transformClassesWithAjxForVariantName`目录下，提供下`buildConfigCache.json`，先确定有没有参与织入。

**Comment by lr784521** [comment]
![759da4ae42840896b88ca6ec3fbad78](https://user-images.githubusercontent.com/36124038/194809392-82941b37-c8ae-4bfa-8118-19e26015fcc6.png)
[buildConfigCache.txt](https://github.com/wurensen/gradle_plugin_android_aspectjx/files/9743708/buildConfigCache.txt)

这是文件  我将  .json  改为  txt  方便上传

**Comment by wurensen** [comment]
看了一下，是有参与到织入的，你先看下`app\build\intermediates\javac\envDevelopRelease\classes\com\huayun\application\page\splash\SplashActivity.class`这个类反编译的情况，方便的话发出来看看。
建议你再随便创建个类，然后针对这个类的某个函数进行织入，测试下是不是正常，因为我发现你这个activity还用到hilt来生成。

**Comment by lr784521** [comment]
这是刚才普通方式设置点击事件编译出来的代码，我项目使用了Hilt 和 阿里 Route
![086bef8eeb09899db730203c3035d9b](https://user-images.githubusercontent.com/36124038/194819257-3b8f7234-d6b4-4710-ba3e-1e9ff38fad62.jpg)
[SplashActivity.zip](https://github.com/wurensen/gradle_plugin_android_aspectjx/files/9744039/SplashActivity.zip)
这是编码后源码。
另外我觉得应该是生效了，原因如下：
当我使用butterknife给控件设置点击事件后,这时候触发点击事件后会崩溃,而用系统原生的ivStart.setOnClickListener设置点击事件 确实没有拦截到，并且不会崩溃。 
资料如下：
![b2cd532081a3ad10bd20fa4d1932919](https://user-images.githu
![58ed703546ee7994f00ea8adb0784ff](https://user-images.githubusercontent.com/36124038/194820154-a8b5935c-8aad-46cf-b166-1a128fb3322f.jpg)
busercontent.com/36124038/194820147-64d7a9ea-6993-4b7d-b5f6-a0ca76e4acce.jpg)
[SplashActivity.zip](https://github.com/wurensen/gradle_plugin_android_aspectjx/files/9744078/SplashActivity.zip)
这是使用butterknife设置点击事件后 的编译源码

**Comment by lr784521** [comment]
这是 使用butterknife设置点击事件，触发事件后崩溃的错误信息
    Process: com.huayun.planet.release, PID: 12348
    java.lang.NoSuchMethodError: No static method aspectOf()Lcom/huayun/application/util/MainUtil; in class Lcom/huayun/application/util/MainUtil; or its super classes (declaration of 'com.huayun.application.util.MainUtil' appears in /data/app/com.huayun.planet.release-FnVQ5LFzkz0fFQ4G6SwVzQ==/base.apk)
        at butterknife.internal.DebouncingOnClickListener.onClick(DebouncingOnClickListener.java:15)
        at android.view.View.performClick(View.java:7257)
        at android.view.View.performClickInternal(View.java:7213)
        at android.view.View.access$3800(View.java:828)
        at android.view.View$PerformClick.run(View.java:27921)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:237)
        at android.app.ActivityThread.main(ActivityThread.java:7830)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1040)

**Comment by wurensen** [comment]
你要清楚这个工具是针对字节码进行修改，所以你的规则也是要匹配上对应的字节码的。
你如果担心织入规则错误，可以用我这个项目的demo来测试，如果我这个有效，证明织入规则是没问题的。

**Comment by lr784521** [comment]
嗯  我试着跑你的源码  没有跑起来，但是 我是复制之前 那个老版本的Aspect 规则确实是没问题，我老版本demo也可以正常运行，然后 现在这个项目主要是用的gradle版本高，老的已经没法用了，所以现在使用了您提供的,配置是按照说明来的，就是现在普通方法拦截 没有生效 ，然后我看规则写的也没有问题，所以我怕是我配置的问题  或者其他方面的疏忽 所以想请您帮忙看下，为什么现在普通方法 没有拦截到

**Comment by lr784521** [comment]
[AspectJ.zip](https://github.com/wurensen/gradle_plugin_android_aspectjx/files/9744232/AspectJ.zip)
这是我之前用老版本写的demo

**Comment by wurensen** [comment]
> 嗯 我试着跑你的源码 没有跑起来，但是 我是复制之前 那个老版本的Aspect 规则确实是没问题，我老版本demo也可以正常运行，然后 现在这个项目主要是用的gradle版本高，老的已经没法用了，所以现在使用了您提供的,配置是按照说明来的，就是现在普通方法拦截 没有生效 ，然后我看规则写的也没有问题，所以我怕是我配置的问题 或者其他方面的疏忽 所以想请您帮忙看下，为什么现在普通方法 没有拦截到

我这个项目没有跑起来？发生什么错误了？你先用我的demo来验证你的规则，这样可以先确定织入规则有没有问题。

**Comment by lr784521** [comment]
这是报错内容
![b3e6d08986b93f0cf5214da8646a9c1](https://user-images.githubusercontent.com/36124038/194825546-ddb996be-cf08-431f-8164-74b638d61822.jpg)

Caused by: groovy.lang.MissingPropertyException: Could not get unknown property 'ossrhUsername' for Credentials [username: null] of type org.gradle.internal.credentials.DefaultPasswordCredentials_Decorated.

是要改maven配置吗？

**Comment by androidKy** [comment]
直接注释掉上传模块的代码就行了

**Comment by wurensen** [comment]
@lr784521 这是我要上传的仓库相关的用户名密码等，你看提示缺少什么key，就自己配置到gradle.properties里面就行，值随便给

**Comment by androidKy** [comment]
> @lr784521 这是我要上传的仓库相关的用户名密码等，你看提示缺少什么key，就自己配置到gradle.properties里面就行，值随便给

兄弟，可以麻烦看下我提的那个问题不

**Comment by lr784521** [comment]
可以了  注释掉 仓库就行  我试了下  你的demo 是正常可以，我再试下我自己编写规则

**Comment by lr784521** [comment]
我试了下  普通的方法是可以拦截到，但是系统的拦截不到了
![0a1a5cc3ef74c5a7b0f2993ac8251ac](https://user-images.githubusercontent.com/36124038/194836455-4fad2a2f-b766-414b-bc25-c583958532f9.jpg)
但是以前老版本AspectJ是可以全局拦截 系统的点击事件的，在我上面提供的demo中有写，并且可以运行，如果您有空的话看一看，能实现老版本的功能是最好的，这样的业务场景比较重要

**Reply by wurensen** [reply]
不可能可以修改到framework的类啊。https://github.com/wurensen/gradle_plugin_android_aspectjx/discussions/43#discussioncomment-3837795 你是说这个用旧版本插件可以，新版本插件不行是吗？

**Comment by lr784521** [comment]
另外 大佬 方便建个  讨论群吗？

**Reply by wurensen** [reply]
我转成discussions了

**Comment by lr784521** [comment]
是的 旧版本可以的，我提供了demo&nbsp; 您可以看下




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "wurensen/gradle_plugin_android_aspectjx"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年10月10日(星期一) 下午5:44
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [wurensen/gradle_plugin_android_aspectjx] 配置完成但拦截不生效 (Discussion #43)





 
不可能可以修改到framework的类啊。#43 (comment) 你是说这个用旧版本插件可以，新版本插件不行是吗？

**Reply by wurensen** [reply]
这个demo升级插件后，是permission织入也无效了，还是仅仅是onclick织入无效？

**Comment by lr784521** [comment]
另外&nbsp; 我自己的项目 配置好了&nbsp; 现在还是拦截不到普通的方法，不知道 是不是差什么？
这是我的配置信息


1.在项目根目录build.gradle添加
&nbsp;classpath 'io.github.wurensen:gradle-android-plugin-aspectjx:3.2.0'


2.在App主工程添加
apply plugin: 'io.github.wurensen.android-aspectjx'


// ajx配置
aspectjx {
&nbsp; &nbsp; enabled = true
&nbsp; &nbsp; // 移除kotlin相关，编译错误和提升速度
&nbsp; &nbsp; exclude 'kotlin.jvm', 'kotlin.internal'
&nbsp; &nbsp; exclude 'kotlinx.coroutines.internal', 'kotlinx.coroutines.android'
}



3.编写测试
@Aspect
public class MainUtil {
&nbsp; &nbsp; @Around("call(* com.huayun.application.page.splash.SplashActivity.test())")
&nbsp; &nbsp; public void setOnClick(ProceedingJoinPoint joinPoint) {
&nbsp; &nbsp; &nbsp; &nbsp; LogUtil.xLoge("setOnClickListenersetOnClickListener");
&nbsp; &nbsp; &nbsp; &nbsp; try {
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; joinPoint.proceed(joinPoint.getArgs());
&nbsp; &nbsp; &nbsp; &nbsp; } catch (Throwable throwable) {
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; throwable.printStackTrace();
&nbsp; &nbsp; &nbsp; &nbsp; }
&nbsp; &nbsp; }

}


这是我的接入流程&nbsp; 不知道是不是少了什么呢？

**Comment by lr784521** [comment]
我就试了 onClick&nbsp; 现在项目里面很需要这个业务




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "wurensen/gradle_plugin_android_aspectjx"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年10月10日(星期一) 晚上6:39
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [wurensen/gradle_plugin_android_aspectjx] 配置完成但拦截不生效 (Discussion #43)





 
这个demo升级插件后，是permission织入也无效了，还是仅仅是onclick织入无效？

**Reply by wurensen** [reply]
你这个demo我试了，插件升级到3.2.0可以正常织入使用（我agp用的4.1.3，gradle用的6.5，jdk用的java11）

**Reply by wurensen** [reply]
还有，你直接reply，不要一直新增comment

**Reply by wurensen** [reply]
<img width="1002" alt="image" src="https://user-images.githubusercontent.com/11225172/194988875-a2fe26f6-3127-4a51-8204-a168296fa777.png">
我在我的demo也试了一下，如果采用lambda表达式来写onClick拦截，会找不到切入点无法正确织入，而采用匿名内部类的写法，是可以正常织入的。
关于lambda如何正确书写表达式，可能得找一找了。

**Reply by wurensen** [reply]
这有别人的提问和暂时的workaround：https://stackoverflow.com/a/31098062/5193118
